sdk/state/update.go

```diff
@@ -143,6 +149,7 @@ func (c *Channel) ConfirmPayment(p *Payment) (payment *Payment, fullySigned bool
 		return p, fullySigned, fmt.Errorf("verifying declaration signed by remote: %w", err)
 	}
 	if !signed {
+		c.latestUnconfirmedPayment = p
```

Comment:
> I think we're missing storing the payment in other cases above. We need to store it once we have signed any transaction for it, because we need to make sure that once we've signed anything for it, we don't ever sign those same types of transactions with different details.

my thinking was that we only want to store the unconfirmedPayment when we return from the method w/o an error. Eg. if for whatever reason signing a transaction returns an error, I didn't think we'd want to store the "bad" input payment that triggered that. Then if a "correct" payment is sent to `ConfirmPayment` because we've stored the "bad" one and the method will fail since the payments don't match.

I added the `defer func`, but to return early if an error was returned
https://github.com/stellar/experimental-payment-channels/pull/81/commits/c887334385256a4a4c1d845e83fb166e455c7c3c

Reply:
> Another thought is by overwriting the entire payment each time there is a chance we lose signatures. It's possible we sign a payment, record the signature, then the payment passed in doesn't have that signature, and then we lose the signature internally. It's important that once we sign a transaction its signature is stored. Maybe we shouldn't set it, but instead merge it. It might be useful if we add a Clone function to Payment too so that we can do deep clones at these points that are well defined. Wdyt?

that's a good point. After thinking about it probably there will be refactoring so I'd like to merge this in and then address that in a separate ticket: https://github.com/stellar/experimental-payment-channels/issues/82